File: kotlin/src/main/kotlin/structural/adapter/MainAdapter.kt
package structural.adapter

internal class OldService {
    fun requestOld(valueOld: String): String {
        return "result with old value = $valueOld"
    }
}

internal interface NewService {
    fun requestNew(valueNew: Int): String
}

internal class AdapterFromNewToOld(private val oldService: OldService) : NewService {
    override fun requestNew(valueNew: Int): String {
        return oldService.requestOld(convertValueFromNewToOld(valueNew))
    }

    private fun convertValueFromNewToOld(valueNew: Int): String {
        return valueNew.toString()
    }
}


fun main(args: Array<String>) {
    val oldService = OldService()
    val adapterFromNewToOld = AdapterFromNewToOld(oldService)
    val result = adapterFromNewToOld.requestNew(11)
    println(result)
}

File: python/structural/adapter/MainAdapter.py
from abc import ABC, abstractmethod

class OldService:
    def requestOld(self, valueOld: str):
        if (isinstance(valueOld, str)):
            return "result with old value = " + valueOld
        else:
            raise ValueError

class NewService(ABC):
    @abstractmethod
    def requestNew(self, valueNew: int):
        pass

class AdapterFromNewToOld(NewService):
    def __init__(self, oldService: OldService):
        if (isinstance(oldService, OldService)):
            self.oldService = oldService
        else:
            raise ValueError

    def requestNew(self, valueNew: int):
        if (isinstance(valueNew, int)):
            return self.oldService.requestOld(self.__convertValueFromNewToOld(valueNew))
        else:
            raise ValueError

    def __convertValueFromNewToOld(self, valueNew: int):
        if (isinstance(valueNew, int)):
            return str(valueNew)
        else:
            raise ValueError

def main():
    oldService = OldService()
    adapterFromNewToOld = AdapterFromNewToOld(oldService)
    result = adapterFromNewToOld.requestNew(11)
    print(result)

main()